R/sim_internal.R

```diff
@@ -0,0 +1,204 @@
+#' Internal simulation function called by the exported simulation functions
+#' within \pkg{simulist}
+#'
+#' @description This internal function simulates a line list, and
+#' when `sim_type` is `"contacts"` or `"outbreak"` a contacts table as well.
+#'
+#' @inheritParams sim_linelist
+#' @inheritParams sim_contacts
+#' @inheritParams .check_sim_input
+#'
+#' @return A `<data.frame>` if `sim_type` is `"linelist"` or `"contacts"`, or a
+#' list of two `<data.frame>`s if `sim_type` is `"outbreak"`.
+#' @keywords internal
+.sim_internal <- function(sim_type = c("linelist", "contacts", "outbreak"),
+                          contact_distribution,
+                          contact_interval,
+                          prob_infect,
+                          onset_to_hosp = NULL,
+                          onset_to_death = NULL,
+                          hosp_risk = NULL,
+                          hosp_death_risk = NULL,
+                          non_hosp_death_risk = NULL,
+                          outbreak_start_date,
+                          add_names = NULL,
+                          add_ct = NULL,
+                          min_outbreak_size,
+                          population_age,
+                          case_type_probs = NULL,
+                          contact_tracing_status_probs = NULL,
+                          config) {
+  sim_type <- match.arg(sim_type)
+
+  outbreak_size <- 0
+  max_iter <- 0L
+  # condition on a minimum chain size
+  while (outbreak_size < min_outbreak_size) {
+    .data <- .sim_network_bp(
+      contact_distribution = contact_distribution,
+      contact_interval = contact_interval,
+      prob_infect = prob_infect,
+      config = config
+    )
+    outbreak_size <- sum(.data$infected == "infected")
+    max_iter <- max_iter + 1L
+    if (max_iter >= 1e3) {
```

Comment:
I would suggest to allow the iterations to be passed down as an argument. It is common to want to adjust these in simulations.

A separate comment is that (in my experience in stats) it is common to call this argument `iter` to specify the amount of iterations. `max_iter` is (as far as I know) an uncommon name - so if you'd like to make it a public argument that could be considered relevant.

Reply:
The `max_iter` variable does not define the number of iterations run, but rather the maximum number of iterations that are permitted given the conditions on the minimum outbreak size (`min_outbreak_size` variable). Therefore I agree it could be nice to let the user increase or decrease `max_iter`, but I disagree about renaming it `iter` as this may falsely imply that this is the number of simulation iterations to run. 

The reasoning for having a `max_iter` is to prevent a very large number of simulations being looped over when the parameters given to the model cannot produce an outbreak with the required minimum size. This prevent the function hanging in a loop.

Let me know what you think and if you agree I can add `max_iter` to the `config` argument to allow advanced users to modify.

Reply:
> The max_iter variable does not define the number of iterations run, but rather the maximum number of iterations that are permitted given the conditions on the minimum outbreak size (min_outbreak_size variable). 

Completely missed that part - projected meaning onto it a bit much then! Thanks for the explainer, wouldn't rename it either in that case. I don't know whether 1000 as a maximum set of iterations is a lot, a little, or okay in this scenario. 

I know that in Monte Carlo simulations 1000 is low-medium, so a max set of iterations would make sense around 10,000 or higher. I don't know how that translates here as it might be a slightly different kind of iteration than what I'm conceiving of. 

All in all, I don't think that this has to be a blocker in any case. If there is a clear reason to have higher `max_iter` or expose it to the user, that need will emerge in due time. Glad we had this exchange so I understand it slightly better - thanks for taking the time to explain 💛

Reply:
In `finalsize::final_size()` this is included in the control list (here, `config`) as `iterations` - a default value is provided that advanced users can modify; so I think @joshwlambert's solution is the right one here

Reply:
Thanks, will leave it as is for now and then, like you say, if the need arises from user requests I'm happy to change this default or expose it to the user.